api/admin/controller/__init__.py

```diff
@@ -2390,43 +2394,55 @@ def _get_collection_protocols(self, provider_apis):
 
         return protocols
 
+    def _get_prior_test_results(self, item, protocol_class=None):
+        # :param item: Either an ExternalSearchIndex or a Collection
+        if hasattr(self, "protocol_class"):
+            protocol_class = self.protocol_class
 
-    def _get_prior_test_results(self, collection, protocolClass):
-        """This helper function returns previous self test results for a given
-        collection if it has a protocol.  Used by both SelfTestsController and
-        CollectionSettingsController
-        """
-        provider_apis = list(self.PROVIDER_APIS)
-        provider_apis.append(OPDSImportMonitor)
+        if not item:
+            return None
 
         self_test_results = None
-        protocol = protocolClass
+        item_type = None
 
-        if not collection or not collection.protocol:
-            return None
+        try:
+            if protocol_class is not None:
+                # We're running self-tests for a collection
+                if not item.protocol or not len(item.protocol):
+                    return None
+                item_type = "collection"
+                provider_apis = list(self.PROVIDER_APIS)
+                provider_apis.append(OPDSImportMonitor)
+
+                if item.protocol == OPDSImportMonitor.PROTOCOL:
+                    protocol_class = OPDSImportMonitor
+
+                if protocol_class in provider_apis and issubclass(protocol_class, HasSelfTests):
+                    if (item.protocol == OPDSImportMonitor.PROTOCOL):
+                        extra_args = (OPDSImporter,)
+                    else:
+                        extra_args = ()
```

Comment:
As you do this work, this code is accumulating special cases -- we do something different if we have an OPDSImportMonitor, we do something different if it's a collection versus a search index.

This isn't something to worry about now, but once you expand this to also cover the self-tests for patron authentication integrations, this code will probably get another `else if` clause, and once that's done it's worth seeing if we can simplify it, maybe by creating specialized subclasses of `HasSelfTests` for collection integrations, patron integrations, and search integrations, and putting the special logic in the subclasses.